### PR TITLE
feat: define iam_policy_statements object syntax

### PR DESCRIPTION
**Pull request by gberenice**
## what

- This updates `iam_policy_statements` variable definition to take Terraform object with optional attributes rather than `type = any`. Provider version is bumped since `optional`, as a non-experimental feature, was introduced in 1.3.

## why

- Provide more control over the expected structure of the input data for IAM policy statements.

## references

- [Slack thread](https://sweetops.slack.com/archives/CUJPCP1K6/p1686000812030009?thread_ts=1684943351.968609&cid=CUJPCP1K6) + [PR #692](https://github.com/cloudposse/terraform-aws-components/pull/692).

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -5,7 +5,27 @@ variable "iam_source_json_url" {
...
+    principals = optional(list(object({
+      type        = string
+      identifiers = list(string)
+    })), [])
+    not_principals = optional(list(object({
+      type        = string
+      identifiers = list(string)
+    })), [])
+  })
   description = "Map of IAM policy statements to use in the policy. This can be used with or instead of the `var.iam_source_json_url`."
```
Comment (Nuru) [inline variables.tf:29]: Change "Map of" to "Object describing"

File: variables.tf [hunk 2]
```diff
@@ -5,7 +5,27 @@ variable "iam_source_json_url" {
...
 
 variable "iam_policy_statements" {
-  type        = any
+  type = object({
+    sid           = optional(string, "")
+    effect        = optional(string, "")
+    actions       = optional(list(string), [])
+    not_actions   = optional(list(string), [])
+    resources     = optional(list(string), [])
+    not_resources = optional(list(string), [])
```
Comment (Nuru) [inline variables.tf:14]: - Add `version` here and in `main.tf`.
- Move statement components to list of objects.
- Inputs that are not going to be expanded into blocks by `dynamic` should default to `null`.
- In `main.tf`, replace lookups like

```hcl
lookup(statement.value, "sid", statement.key)
```

With direct references

```hcl
statement.value.sid
```

```suggestion
  type = object({
    version       = optional(string, null)
    statement = list(object({
      sid           = optional(string, null)
      effect        = optional(string, null)
      actions       = optional(list(string), null)
      not_actions   = optional(list(string), null)
      resources     = optional(list(string), null)
      not_resources = optional(list(string), null)
    })), [])
```

**Review by Nuru — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 2 above)
Comment (gberenice) [reply variables.tf]: @Nuru thanks a lot for your review!
* Version arguments should be set for data source `aws_iam_policy_document`, so I added it as a separate variable. 
* I updated statements to the map of objects - that's what is expected [in the example](https://github.com/cloudposse/terraform-aws-iam-policy/blob/main/examples/complete/fixtures.us-east-2.tfvars#L58).
* Inputs were updated.
* Lookups were replaced👍 

Please let me know your thoughts.

Comment (Nuru) [reply variables.tf]: @gberenice I forgot we want this to be backward compatible and that means the input is a map. I will take it from here. Thank you very much for your contributions.

Comment (gberenice) [reply variables.tf]: @Nuru sounds good, thank you!

**Pull request merged**